web/source/kmwnative.js

```diff
@@ -94,22 +94,30 @@
   * 
   * @param  {Object}  Ln  keyboard stub object    
   */    
-  keymanweb.keyboardUnavailable = function(Ln)
-  {
-    return window.setTimeout(function() {
-        util.wait(false);
-        var Ps=keymanweb._KeyboardStubs[Ln],kbdName=Ps['KN'],lgName=Ps['KL'];
-        kbdName=kbdName.replace(/\s*keyboard\s*/i,'');
+  keymanweb.keyboardUnavailable = function(Ln) {
+    keymanweb.loadFailureHandler = function(altString) {
```

Comment:
OK. I understand now what this code is doing. I like the changes you've made, but .... 

The existing code that you have had to change is pretty horrible, and I think we should take the opportunity to refactor it further, and try and eliminate  `loadTimer`, `loadFailureHandler`, `_LoadingInternalName` as globals. I think that the very out-of-sequence `KR` references to `loadTimer` and `_LoadingInternalName` are also problematic: `KR`'s concern should be registering the keyboard, and keyboard activation, if wanted, should be handled by the `onload` event. That's far less side-effecty than the current situation.

So, can we look at something like the following:

1. rename `keyboardUnavailable` to e.g. `setupKeyboardLoadFailureHandler` (can you think of a better name?)

2. change `keyboardUnavailable` (aka new-improved-name) to return an object that includes a timer and the callback:
````
var result = { callback: function(altString) { /* loadFailureHandler content */ } };
result.timer = window.setTimeout(result.callback, 10000);
return result;
````

3. Then we change the scope of `keymanweb.loadTimer`, moving it to the `kbdStub.loadingState`, which encapsulates `timer` and `callback` (and removes the need for _LoadingInternalName).

4. With this done, I think we can refactor the references to `loadTimer`, `loadFailureHandler`, `_LoadingInternalName`. I think with a little bit more work, we can eliminate this global state altogether, and not break the loading of multiple keyboards in a row, which is currently fragile and uncomfortable. The cleanup of timer, etc, should be done in the `onerror` and `onload` events only, and never in `KR`. `KR` should not need to be aware of this; in fact, keyboard activation should be done in the `onload` event.

That's a lot more to work through. Would be keen to discuss with you when you get a chance.